Integrated syzkaller with mac80211_hwsim

a-nogikh opened the pull request:
Two virtual wireless devices are instantiated during network devices initialization.
Also, two pseudo syscalls are added:
1) syz_80211_inject_frame(mac_addr, packet, packet_len) - injects an arbitrary packet into the wireless stack. It is injected as if it originated from the device identitied by mac_addr. 
2) syz_80211_join_ibss(ssid, ssid_len, fix_requency, await_up) - sets up and configures an IBSS network. If await_up=1, does not return until the network is established

This code reuses netlink helpers that were already present in the `executor/linux_common.h` file. This required some minor changes, which do not break the interface of these helpers. Details can be found in commit descriptions.

Also, this pull request contains minor modifications to nl80211 descriptions that focus fuzzer's attention on virtual devices and interfaces introduced here.

melver commented on executor/common_linux.h:
```diff
@@ -469,7 +470,7 @@ static void netlink_add_neigh(struct nlmsg* nlmsg, int sock, const char* name,
 #endif
 
-#if SYZ_EXECUTOR || SYZ_NET_DEVICES || SYZ_NET_INJECTION
+#if SYZ_EXECUTOR || SYZ_NET_DEVICES || SYZ_NET_INJECTION || SYZ_DEVLINK_PCI
```
Can you fold the changes from "common/common_linux.h: addressed CI errors" into the relevant commits (i.e. amend with rebase -i etc.)?

a-nogikh replied:
Done

melver commented on executor/common_linux.h:
```diff
@@ -206,6 +206,37 @@ static int netlink_send(struct nlmsg* nlmsg, int sock)
 	return netlink_send_ext(nlmsg, sock, 0, NULL);
 }
 
+// TODO: switch other methods to netlink_query_family_id
```
Will this be resolved in this PR?

a-nogikh replied:
It was not planned to resolve it in this PR. This PR focuses on changes that are necessary to fuzz 802.11, while this TODO is about general refactoring that will additionally require testing of the affected subsystems.

melver replied:
Hmm, I see 2 almost identical functions: netlink_wireguard_id_get, netlink_devlink_id_get.

If all these functions are essentially identical, I don't see any harm in adding a commit at the end of this PR that replaces uses of those with your more generic function that just get passed the right parameter.

Some things to keep in mind:
- There might be some careful #if needed to make sure the generic function is compiled for the configs where we need it.
- Maybe make the error messages print the family_name, as that's what seem to be the main difference between them.

Also, what would make the whole thing more reviewable is simply having another preparatory commit before this commit that simply does the refactor and adds this generic function. That way, it should be obvious from the commit that the functions being replaced are identical with the generic one.

If it turns out it's not that trivial and I missed something, then feel free to leave alone.

melver commented on executor/common_linux.h:
```diff
@@ -1025,6 +1086,41 @@ static void netlink_wireguard_setup(void)
+	return 0;
+}
+
+// TODO: initialize this only when HWSIM is present
```
Could this be done in this PR? What is needed to check if HWSIM is present?

melver commented on executor/common_linux.h:
```diff
@@ -722,6 +757,32 @@ static void initialize_devlink_pci(void)
+#define HWSIM_CMD_REGISTER 1
+#define HWSIM_CMD_FRAME 2
+#define HWSIM_CMD_NEW_RADIO 4
+#define HWSIM_CMD_DEL_RADIO 5
```
I think DEL and GET isn't used yet. In case it'll be used in a follow-up PR it could possibly stay, but prefer removing them.

a-nogikh replied:
Removed these extra definitions.

melver commented on executor/common_linux.h:
```diff
@@ -1025,6 +1086,41 @@ static void netlink_wireguard_setup(void)
+// TODO: initialize this only when HWSIM is present
+static void initialize_hwsim80211(void)
+{
+	static uint8 mac_addr[6] = NETDEV_80211_MAC_BASE;
```
Does this need to be static?

a-nogikh replied:
No, it doesn't. Fixed.

melver commented on executor/common_linux.h:
```diff
@@ -128,11 +129,9 @@ struct nlmsg {
+	char buf[4096];
 };
 
-static struct nlmsg nlmsg;
```
Does this need to be moved in this commit?

Could it be a preparatory/cleanup commit?

a-nogikh replied:
Created a separate commit.

melver commented on executor/common_linux.h:
```diff
@@ -128,11 +129,9 @@ struct nlmsg {
 	int nesting;
 	struct nlattr* nested[8];
-	char buf[1024];
+	char buf[4096];
```
This increase could probably also be a separate commit with an explanation that it's a preparatory commit for WiFi fuzzing.

a-nogikh replied:
Created a separate commit.

melver commented on executor/common_linux.h:
```diff
@@ -150,7 +149,8 @@ static void netlink_attr(struct nlmsg* nlmsg, int typ,
 	attr->nla_len = sizeof(*attr) + size;
 	attr->nla_type = typ;
-	memcpy(attr + 1, data, size);
+	if (size > 0)
```
I think this as well could be a preparatory commit, just something like "allow zero-payload attribute changes ...."?

a-nogikh replied:
Created a separate commit.

melver commented on executor/common_linux.h:
```diff
@@ -206,6 +206,37 @@ static int netlink_send(struct nlmsg* nlmsg, int sock)
+	genlhdr.cmd = CTRL_CMD_GETFAMILY;
+	netlink_init(nlmsg, GENL_ID_CTRL, 0, &genlhdr, sizeof(genlhdr));
+	netlink_attr(nlmsg, CTRL_ATTR_FAMILY_NAME, family_name, strnlen(family_name, GENL_NAMSIZ - 1) + 1);
+	int n = 0;
```
If you do the refactor to switch to this generic one, I'd really make the generic one as identical to the ones we already have as possible, that includes leaving the variable declarations as-is (unless this 'n' needs to be initialized to 0) and at the start of the function. Essentially we should just be able to put them side-by-side and see no diff except the family_name.

a-nogikh replied:
The new generic function is idential to the functions it replaces.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -4564,3 +4656,236 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+		nl80211_join_ibss(&tmp_msg, sock, nl80211_family_id, interface, &ibss_props);
+
+		if (await_up) {
+			while (get_ifla_operstate(&tmp_msg, interface) != IF_OPER_UP)
```
It waits ~ 9 seconds if wiphy_freq_fixed is false, and less than 0.5 seconds otherwise.

Definitions in `sys/linux/net_80211.txt` already take account of that.

melver commented on executor/common_linux.h:
```diff
@@ -1094,8 +1094,10 @@ static int hwsim80211_create_device(int sock, int hwsim_family, uint8 mac_addr[E
 	int err = netlink_send(&msg, sock);
-	// TODO: in case of success hwsim returns index instead of 0. Fix it.
-	if (err < 0) {
+	// TODO: in case of success, hwsim returns index instead of 0, while netlink_send_ext supposes
```
(I think this commit should be folded into the original commit that introduced that change. If things are still WIP, please ignore this comment.)

a-nogikh replied:
Yes, at the moment thigns are temporarily WIP. I'll squash commits after I complete work

a-nogikh commented on executor/common_linux.h:
```diff
@@ -205,6 +206,37 @@ static int netlink_send(struct nlmsg* nlmsg, int sock)
 	return netlink_send_ext(nlmsg, sock, 0, NULL);
 }
 
+// TODO: replace other similar methods by netlink_query_family_id
```
netlink_devlink_id_get and netlink_wireguard_id_get methods are doing essentially the same thing, while this function was a generic replacement for them. This TODO is now fixed by a separate commit as part of this pull request.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -1028,6 +1084,41 @@ static void netlink_wireguard_setup(void)
+	netlink_attr(&msg, HWSIM_ATTR_SUPPORT_P2P_DEVICE, NULL, 0);
+	netlink_attr(&msg, HWSIM_ATTR_PERM_ADDR, mac_addr, ETH_ALEN);
+	int err = netlink_send(&msg, sock);
+	// TODO: in case of success hwsim returns index instead of 0. Fix it.
```
Expanded the TODO note

a-nogikh commented on sys/linux/net_80211.txt:
```diff
@@ -0,0 +1,33 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+type _80211_mac_addr[LAST] {
```
I used `_` to denote that the type is a helper that is only needed to define other types. 

Renamed it.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -4564,3 +4656,236 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+	}
+
+	int hwsim_family_id = netlink_query_family_id(&tmp_msg, sock, "MAC80211_HWSIM");
+	hwsim_register_socket(&tmp_msg, sock, hwsim_family_id);
```
Updated this and similar code to handle bad return values.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -4564,3 +4656,236 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+		set_interface_state(interface, 0);
+		nl80211_set_interface(&tmp_msg, sock, nl80211_family_id, interface, NL80211_IFTYPE_ADHOC);
+		set_interface_state(interface, 1);
+		nl80211_join_ibss(&tmp_msg, sock, nl80211_family_id, interface, &ibss_props);
```
Updated this and similar code to handle bad return values.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -4564,3 +4656,236 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+	netlink_attr(nlmsg, NL80211_ATTR_IFINDEX, &ifindex, sizeof(uint32));
+	netlink_attr(nlmsg, NL80211_ATTR_SSID, props->ssid, props->ssid_len);
+	netlink_attr(nlmsg, NL80211_ATTR_WIPHY_FREQ, &(props->wiphy_freq), sizeof(uint32));
+	if (props->mac) {
```
Removed braces where it seemed appropriate.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -1028,6 +1084,41 @@ static void netlink_wireguard_setup(void)
+	int hwsim_family_id = netlink_query_family_id(&nlmsg, sock, "MAC80211_HWSIM");
+	for (int device_id = 0; device_id < NETDEV_80211_INTITIAL_DEVICE_COUNT; device_id++) {
+		mac_addr[5] = device_id;
+		hwsim80211_create_device(sock, hwsim_family_id, mac_addr);
```
Updated this and similar code to handle bad return values.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -1028,6 +1084,41 @@ static void netlink_wireguard_setup(void)
+	uint8 mac_addr[6] = NETDEV_80211_MAC_BASE;
+	int sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_GENERIC);
+	if (sock < 0) {
+		debug("initialize_80211hwsim: failed to create socket (%d)\n", errno);
```
Fixed this and similar issues.

melver reviewed (approved):
Great, thank you!

Please wait for Andrey's and Dmitry's feedback.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -107,14 +107,15 @@ static bool write_file(const char* file, const char* what, ...)
 #endif
 
-#if SYZ_EXECUTOR || SYZ_NET_DEVICES || SYZ_NET_INJECTION || SYZ_DEVLINK_PCI
+#if SYZ_EXECUTOR || SYZ_NET_DEVICES || SYZ_NET_INJECTION || SYZ_DEVLINK_PCI || __NR_syz_80211_join_ibss || __NR_syz_80211_inject_frame
```
I've added a new flag (and the corresponding stuff) to control wifi fuzzing. See latest commit.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -725,6 +757,30 @@ static void initialize_devlink_pci(void)
 #endif
 #endif
 
+#if __NR_syz_80211_join_ibss || __NR_syz_80211_inject_frame || SYZ_EXECUTOR || SYZ_NET_DEVICES
```
Done. 

Unfortunately, such replacement was only possible in this specific line. I haven't found a way to tell syzkaller never to generate a C program with `__NR_syz_80211_join_ibss` or `__NR_syz_80211_inject_frame`, but without `SYZ_WIFI`. There is [one mechanism](https://github.com/google/syzkaller/blob/master/pkg/csource/csource.go#L210), but it doesn't seem to prevent `__NR*` defines, and this leads to warnings about unused functions.

xairy commented on executor/common_linux.h:
```diff
@@ -107,7 +107,7 @@ static bool write_file(const char* file, const char* what, ...)
 #endif
 
-#if SYZ_EXECUTOR || SYZ_NET_DEVICES || SYZ_NET_INJECTION || SYZ_DEVLINK_PCI || __NR_syz_80211_join_ibss || __NR_syz_80211_inject_frame
+#if SYZ_EXECUTOR || SYZ_NET_DEVICES || SYZ_NET_INJECTION || SYZ_DEVLINK_PCI || SYZ_WIFI || __NR_syz_80211_inject_frame || __NR_syz_80211_join_ibss
```
Please squash this commit with the one that adds the pseudo-syscalls, otherwise it's hard to review.

xairy replied:
Or put the config option commit first with empty implementation of pseudo-syscalls, and then fill in the implementation in another commit.

xairy commented on executor/common_linux.h:
```diff
@@ -726,6 +726,75 @@ static void initialize_devlink_pci(void)
+	netlink_attr(&msg, HWSIM_ATTR_SUPPORT_P2P_DEVICE, NULL, 0);
+	netlink_attr(&msg, HWSIM_ATTR_PERM_ADDR, mac_addr, ETH_ALEN);
+	int err = netlink_send(&msg, sock);
+	// TODO: in case of success, hwsim returns index instead of 0, while netlink_send_ext supposes
```
Is this hard to fix?

a-nogikh replied:
Resolved this TODO in a separate commit

xairy commented on executor/common_linux.h:
```diff
@@ -4535,3 +4613,280 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+	return -1;
+}
+
+static long syz_80211_join_ibss(volatile long a0, volatile long a1, volatile long a2, volatile long a3)
```
This pseudo-syscall still seems suspicious to me. Would it make sense to change it to only make one device join the network instead of all of them?

xairy commented on executor/common_linux.h:
```diff
@@ -4535,3 +4613,280 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+			goto error;
+		}
+
+		if (await_up) {
```
Hm, can we ask all devices to join the network first, and then wait until they do? Instead of doing it one by one.

a-nogikh replied:
I'm worried that this additional flexibility can complicate generation of good progs. I've implemented it in a separate branch  and starter a separate syzkaller instance. If it works fine, I'll push that commit into this pull request.

a-nogikh replied:
Syzkaller from this pull request detected 12 types of crashes during 3 days, while syzkaller with `syz_80211_join_ibss(wlan_interface, ssid, ssid_len, fix_requency, await_up)`  detected only 7 types of crashes during the same time period. Of course, this can also be attributed to (un)luck as the process is somewhat random, but still the difference is sufficient.

a-nogikh replied:
Oh, I posted my replies into the wrong comment. They were in repsonse to `This pseudo-syscall still seems suspicious to me. Would it make sense to change it to only make one device join the network instead of all of them?`

xairy reviewed (commented):
Nit: use present tense in commit titles, not past.

a-nogikh commented on pkg/csource/generated.go:
```diff
@@ -2354,11 +2354,9 @@ struct nlmsg {
 	int nesting;
 	struct nlattr* nested[8];
-	char buf[1024];
+	char buf[4096];
```
Fixed this and similar issues

dvyukov commented on executor/common_linux.h:
```diff
@@ -205,6 +206,36 @@ static int netlink_send(struct nlmsg* nlmsg, int sock)
 	return netlink_send_ext(nlmsg, sock, 0, NULL);
 }
 
+static int netlink_query_family_id(struct nlmsg* nlmsg, int sock, const char* family_name)
```
I think syz_genetlink_get_family_id pseudo-syscall also doing the same, so potentially can be refactored to use this function as well.

a-nogikh replied:
Refactored that function in a separate commit

dvyukov commented on sys/linux/socket_netlink_generic_80211.txt:
```diff
@@ -11,12 +11,16 @@ include <linux/ieee80211.h>
 type nl80211_whipy int32[-1:4]
 
 syz_genetlink_get_family_id$nl80211(name ptr[in, string["nl80211"]]) nl80211_family_id
+ioctl$sock_SIOCGIFINDEX_80211(fd sock, cmd const[SIOCGIFINDEX], arg ptr[inout, ifreq_dev_t[nl80211_devnames, nl80211_ifindex[opt]]])
```
Not directly related to your change as we do it this way everywhere. But @necipfazil recently added field direction to structs, and this looks like a perfect use case for them. We declare these ifindex'es as opt only because they are actually out, while the ifr_ifrn field in ifreq is in. We could specify directions directly now.

a-nogikh replied:
Adjusted `ifreq_dev_t` and its uses in a separate commit

dvyukov commented on executor/common_linux.h:
```diff
@@ -4535,3 +4610,280 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+	}
+
+	int hwsim_family_id = netlink_query_family_id(&tmp_msg, sock, "MAC80211_HWSIM");
+	int ret = hwsim_register_socket(&tmp_msg, sock, hwsim_family_id);
```
This registers netlink port id somewhere in hswim subsystem. But we always use 0 port id. Can't we do this once in initialize_wifi_devices?

a-nogikh replied:
No, we cannot do that because this command registers a specific socket. Unregistration happens automatically once the socket is closed.

Also, we cannot create one socket, register it, and use all the time, because this prevents mac80211_hwsim from emulating perfect medium. It disrupts normal networking activities.

dvyukov commented:
Couldn't we move what we do in syz_80211_join_ibss and maybe even hwsim_register_socket in syz_80211_inject_frame into initialize_wifi_devices? That's how we do it for all other devices, e.g. IFF_UP happens when we setup net devices. It would speed up fuzzing and provide devices in some meaningful/useful state right away. Nominally creating devices if they are not ready to actual data exchange is not that useful.
Also it can make bugs more reproducible. Currently all operations on the devices (syz_80211_join_ibss/hwsim_register_socket) alter state of the devices and this sinks to subsequent tests. So if one test does syz_80211_join_ibss and next does syz_80211_inject_frame, it will work. But reproducing crash in syz_80211_inject_frame won't work, because the test is not hermetic.

dvyukov commented:
Are there ways to make these devices interact more with descriptions in sys/linux/socket_netlink_generic_80211.txt? We should make descriptions in socket_netlink_generic_80211.txt use our new device names, ifindex'es and mac addresses. But maybe there are more ways to make them interact.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -4535,3 +4610,280 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+	struct nlmsg tmp_msg;
+
+	if (buf == NULL || buf_len < 0 || buf_len > WIFI_MAX_INJECT_LEN)
+		return -1; // restricting input in order to avoid unnecessary segfaults on memcpy
```
Removed the check of buf pointer.

a-nogikh commented:
@dvyukov 

I addressed all your comments. 
For reasons specified blow, code review fixes come in two commits and are not squashed yet. 
`all: apply code review fixes` will be squashed with `all: integrate with mac80211_hwsim`, 
`all: introduce an enhanced version of syz_80211_join_ibss` will be either squashed with `all: integrate with mac80211_hwsim` or deleted.

> Couldn't we move what we do in syz_80211_join_ibss and maybe even hwsim_register_socket in syz_80211_inject_frame into initialize_wifi_devices? 

I moved IBSS network creation to `initialize_wifi_devices` and removed the `syz_80211_join_ibss` pseudo syscall. Interestingly, local syzkaller executions (lasting 2-3 days) demonstrated that the number of detected issues has significantly decreased. Or at least it became noticeably harder to hit them.

I even implemented exact descriptions for NL80211_CMD_SET_INTERFACE and NL80211_CMD_JOIN_IBSS (i.e. not the kind of netlink descriptions used right now - with array of union of attribute types, but as a firm structure). It also didn't seem to help much. It seems that `syz_80211_join_ibss` really has some positive effect on generating problematic situations.

So I introduced it again, but in a slightly different form
1) It accepts a network interface parameter (as suggested by @xairy)
2) It does not attempt to execute for a long time. It either starts a long procedure and exits, starts a short procedure and exists, or starts a short procedure and waits until it's complete.

What do you think about it? Should we have this pseudo call or only rely on `initialize_wifi_devices` and syzkaller descriptions?
It is added as a separate commit so that it'll be easy to remove it if necessary.

a-nogikh commented:
> Are there ways to make these devices interact more with descriptions in sys/linux/socket_netlink_generic_80211.txt? We should make descriptions in socket_netlink_generic_80211.txt use our new device names, ifindex'es and mac addresses. But maybe there are more ways to make them interact.

`socket_netlink_generic_80211.txt` has already been updated to use the same device names, ifindexes, mac addressess, ssid's and bssid's as those that are used to create virtual devices.

a-nogikh commented on sys/linux/net_80211.txt:
```diff
@@ -0,0 +1,27 @@
+
+type ieee80211_frame array[int8]
+
+syz_80211_join_ibss(ssid ptr[in, ieee80211_ssid], ssid_len len[ssid], freq_fixed int32[0:1], await_up int32[0:1]) (timeout[9500], prog_timeout[11000])
```
It was performing a channel scan. The updated version of the pseudo syscall does not wait that much anymore

a-nogikh commented on executor/common_linux.h:
```diff
@@ -726,6 +726,72 @@ static void initialize_devlink_pci(void)
+	int hwsim_family_id = netlink_query_family_id(&nlmsg, sock, "MAC80211_HWSIM");
+	for (int device_id = 0; device_id < WIFI_INITIAL_DEVICE_COUNT; device_id++) {
+		mac_addr[5] = device_id;
+		int ret = hwsim80211_create_device(sock, hwsim_family_id, mac_addr);
```
It's implicit. Added a comment.

dvyukov commented:
Nice! This ifreq change and socket_netlink_generic_80211.txt changes make things better.

dvyukov commented on executor/common_linux.h:
```diff
@@ -181,30 +182,71 @@ static int netlink_send_ext(struct nlmsg* nlmsg, int sock,
-	if (n != hdr->nlmsg_len)
-		fail("short netlink write: %d/%d", n, hdr->nlmsg_len);
+	if (n != hdr->nlmsg_len) {
+		debug("short netlink write: %d/%d", n, hdr->nlmsg_len);
```
Can this (and subsequent ones) legitimately fail now?
We used `fail` here for a reason. This is not exactly normal production C code. Because this is fuzzer-driven lots of errors are silently ignored (and some callers of this function don't even bother checking return code). So if something fails unexpectedly (a bug in executor code), it won't be noticed. We will just get some drop in coverage which nobody will notice.
If there is a fail, it will be noticed as it will crash whole VMs. And since we had fail here for some time now and proved to not fail all the time (one of few bits here that one can really rely on now), it's pity to start giving up ground.

a-nogikh replied:
If fault injections are enabled, VMs begin to constantly crash as `lost connection to test machine`. So we either need to avoid the use of these netlink helpers for pseudo syscalls (which will bloat code size), or make them fail gracefully. 

I re-implemented this patch, now it is possible to specify whether the function should handle an error gracefully or not. The implementation is a bit hacky in order to keep the function short and avoid duplication.

dvyukov replied:
Ouch. Good find. Frankly, I never considered such a possibility. But it looks very realistic.

However, the current approach you took does not look very scalable. This problem also seems to affect more functions. syz_init_net_socket uses fail and it can perfectly fail with fault injection. read_tun -- the same (maybe we even observed this, there is some strange logic for errno). And I assume more functions are subject to this. With your approach we would need to carefully examine code, explicitly opt-in each function and then carefully use the graceful flag. Any wrong value for graceful will lead to the same problem, but as we see it's quite hard to notice, so most likely these bugs will stay there for a very long time.

I wonder if we could change fail instead. If fault injection is enabled, it could exit with 0 status like exitf. This will solve the problem once and for all, and will also be a good place for a comment.
To check if fault injection is enabled right now for the current thread, we will need to add some __thread variable somewhere. We already have current_cover, maybe we should have current_thread instead with all the data regarding current thread. I can't say what exactly is the right way to organize it, it needs some experimentation to find a nice solution that works.

a-nogikh replied:
That would indeed cover all benign failures. But do we really want to terminate execution in such cases? Technically, nothing prevents the fuzzer from running other syscalls there. And it might even be desirable to continue execution as further syscalls might leverage the mess, which was possibly created by failure injection, and hit more intricate kernel bugs. Exiting can cause the kernel to clean things up and some kernel bugs may just go unnoticed.

dvyukov replied:
My reasoning was two-fold:
1. I don't see a nice way to achieve this. If we make fail return in some cases, we will need to change all call sites to handle return. And it will be somewhat confusing semantics for a fail function, and it will be exercised in very rare cases, so developers won't even test these scenarios. People also get used to fail being no-return, so we will sure get bugs in future where we don't handle return case. And for e.g. syz_init_net_socket there is no way out, we need to restore the test namespace.
2. I don't think it gives too much interesting coverage. It only affects pseudo-syscalls and it can achieve any new coverage it gets here with other means (normal syscalls). It will only make few programs where we inject failures into particular points in particular pseudo-syscalls exit early, which seems fine. Programs exit prematurely all the time (e.g. if the test contains exit_group).

a-nogikh replied:
Thank you for the detailed explanation.

I implemented it as you suggested - now there's a thread-local variable that is used to access information about the thread itself. I tested the patch locally, now syzkaller does not generate regular `lost connection to test machine`  crashes when fault injection is enabled.

xairy commented on executor/common_linux.h:
```diff
@@ -722,6 +738,278 @@ static void initialize_devlink_pci(void)
+
+static void initialize_wifi_devices(void)
+{
+	// Set up virtual wifi devices and join them into an IBSS network
```
`.` at the end.

a-nogikh commented:
Fixed new review comments by @xairy and squashed all commits with CR fixes into the main commit.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -171,8 +172,21 @@ static void netlink_done(struct nlmsg* nlmsg)
-			    uint16 reply_type, int* reply_len)
+			    uint16 reply_type, int* reply_len, bool graceful)
 {
+	void (*error)(const char*, ...) = (typeof(error))printf;
```
I used it as a harmless substitute for debug and error. Maybe it's indeed unnecessary. 

Replaced by a NULL value and a series of if checks.

melver replied:
Ah sorry, that makes it worse I think.

Essentially before all this it was 'fatal'. Can we not use 'fatal' or 'debug' here as a default? If not, then 'printf' is perfectly reasonable.

a-nogikh commented:
PTAL

dvyukov commented:
Excited to observe syzbot reaction!
And thanks for fixing the fault injection/fail, this will benefit fuzzing overall.
